ssl/ssl_aead_ctx.cc

```diff
@@ -428,4 +429,79 @@ bool SSLAEADContext::GetIV(const uint8_t **out_iv, size_t *out_iv_len) const {
          EVP_AEAD_CTX_get_iv(ctx_.get(), out_iv, out_iv_len);
 }
 
+#define SSLAEADCONTEXT_SERDE_VERSION 1
+
+static const unsigned int kSSLAEADContextCipherStateTag =
+    CBS_ASN1_CONSTRUCTED | CBS_ASN1_CONTEXT_SPECIFIC | 0;
+
+/*
+ * SSLAEADContextVersion ::= INTEGER {v1 (1)}
+ *
+ * SSLAEADContext ::= SEQUENCE {
+ *   serializationVersion SSLAEADContextVersion,
+ *   cipher         INTEGER,
+ *   cipherState    [0] OCTET STRING
+ * }
+ */
+int SSLAEADContext::SerializeState(CBB *cbb) const {
+  uint32_t cipher_id = SSL_CIPHER_get_id(cipher_);
+
+  CBB seq;
+  CBB cipher_state_octet;
+  ScopedCBB cipher_state;
+
+  if (!CBB_add_asn1(cbb, &seq, CBS_ASN1_SEQUENCE) ||
+      !CBB_add_asn1_uint64(&seq, SSLAEADCONTEXT_SERDE_VERSION) ||
+      !CBB_add_asn1_uint64(&seq, cipher_id)) {
+    OPENSSL_PUT_ERROR(SSL, ERR_R_MALLOC_FAILURE);
+    return 0;
+  }
+
+  CBB_init(cipher_state.get(), 1024);
+  if (!CBB_add_asn1(&seq, &cipher_state_octet, kSSLAEADContextCipherStateTag)) {
+    OPENSSL_PUT_ERROR(SSL, ERR_R_MALLOC_FAILURE);
+    return 0;
+  }
+
+  if (!EVP_AEAD_CTX_serialize_state(ctx_.get(), cipher_state.get()) ||
+      !CBB_add_asn1_octet_string(&cipher_state_octet,
```

Comment:
So what is happening here is encapsulation. For SSLAEADContext it will write out the cipher state as an opaque OCTET STRING whose content just so happens to contain an ASN.1 encoded value. That ASN.1 encoded value, in the AES GCM case, would be a ASN.1 sequence. This is an idea I took from how PKCS#8 does encoding, as it allows us to leave the underlying EVP_AEAD_CTX free to pick how it wants to serialize it's state.

So first I was attaching the tag into the top-level sequence and get a handle to write out the value for that tag, then constructing a new CBB and allow the underlying cipher to serialize it's state to that buffer, then I take the data within that CBB and write it back to the handle I got for the tag and write it out as an OCTET STRING.